redbot/core/bot.py

```diff
@@ -150,6 +151,63 @@ def __init__(self, *args, cli_flags=None, bot_dir: Path = Path.cwd(), **kwargs):
 
         self._permissions_hooks: List[commands.CheckPredicate] = []
         self._red_ready = asyncio.Event()
+        self._red_before_invoke_objs: Set[Coroutine] = set()
+
+    @property
+    def _before_invoke(self):
+        return self._red_before_invoke_method
+
+    @_before_invoke.setter
+    def _before_invoke(self, val):
+        """Prevent this from being overwriten in super().__init__"""
+        pass
+
+    async def _red_before_invoke_method(self, ctx):
+        await self.wait_until_red_ready()
+        return_exceptions = not isinstance(ctx.command, commands.commands._AlwaysAvailableCommand)
+        await asyncio.gather(
+            *(coro(ctx) for coro in self._red_before_invoke_objs),
+            return_exceptions=return_exceptions,
+        )
+
+    def remove_before_invoke_hook(self, coro: Coroutine):
+        """
+        Functional method to remove a before_invoke hooks
+        """
+        self._red_before_invoke_objs.discard(coro)
+
+    def before_invoke(self, coro: T_BIC) -> T_BIC:
+        """
+        Overriden Decorator method for Red's before_invoke behavior
+        
+        This can safely be used purely functionally as well.
```

Comment:
Does this sentence mean that I don't have to use it as deco or...?

Reply:
Yes. `bot.before_invoke(cog.some_func)` is acceptable.